cmake-edit-core/src/main/java/ru/urururu/cmakeedit/core/parser/ExpressionParser.java

```diff
@@ -70,28 +69,43 @@ public static ExpressionNode parseExpression(ParseContext ctx) throws ParseExcep
                 throw new ParseException(ctx, "Key " + key + " is not used yet. For now only $ENV{..} is allowed");
             }
 
-            sb.append(keyBuilder);
-
             if (ctx.reachedEnd()) {
                 throw new ParseException(ctx, "Unexpected end of source");
             }
 
             char closingBracket = BRACKETS.get(ctx.peek());
 
-            sb.append(ctx.peek());
             ctx.advance();
 
+            StringBuilder sb = new StringBuilder();
+            SourceRef constantStart = ctx.position();
+            SourceRef constantEnd = null;
+
             while (!ctx.reachedEnd()) {
                 char cur = ctx.peek();
                 if (cur == closingBracket) {
-                    sb.append(cur);
-                    return new ExpressionNode(key, sb.toString(), nested, start, ctx.position());
+                    if (sb.length() != 0) {
+                        nested.add(new ConstantNode(sb.toString(), constantStart, constantEnd));
+                    }
+
+                    return new ExpressionNode(key, nested, start, ctx.position());
                 } else if (cur == '$') {
+                    if (sb.length() != 0) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS134)